cylc/cylc-flow
# test fixes

Miscellaneous test fixes and stability improvements to fix issues encountered when running the fully expanded test matrix against non-containerised platforms.

Test changes:

* Fix test platform configuration errors.
* Fix tests overly reliant on precise timings or extend timeouts where required (e.g. at submission should be expected to take longer than background).
* Fix lsof filesystem polling bug in the `test_header` (mb).
* Fix fork-bomb test - tests/flakyfunctional/job-submission/18-check-chunking.t
* Replace polling with event-driven stuff where possible.

Non-test changes:

* Re-implement the bash `cylc graph-diff` command in Python as `cylc graph --diff`.
  * The way the workflow argument is handled by the old script was different to the Python scripts.
  * The use of `cylc scan` made it fragile on test platforms.
  * This was the last bash command which allows us to remove the back-compat `os.exec` logic.
  * Change `cylc graph` to use the standard workflow argument parsing function.
  * Note: this drops support for graphical diff tools :( can re-implement later if desired.
* Change the test battery base registration.
  * Ensure that each registration is unique so that test runners on different platforms don't fight over the same files.
  * Make the base reg configurable to make automated testing easier.
* Add a new platform configuration for the maximum number of jobs permitted in a single batch submission.
  * Not expecting this configuration to be used, it just replaces a hard-coded default making testing easier.

Some tests now run faster, cc https://github.com/cylc/cylc-flow/issues/3046

**Requirements check-list**
- [ ] Does not contain off-topic changes (off-topic changes listed above)
- [x] Already covered by existing tests.
- [ ] Appropriate change log entry included.
- [x] No documentation update required.
